### PR TITLE
test: improve foundry test with best practices

### PR DESCRIPTION
**Pull request by zoruka**
Example after running the test suits:
```
➜  non-fungible-apps git:(feature/sc-356/improve-foundry-test-with-best-practices) ✗ yarn test:foundry
yarn run v1.22.19
$ forge test --via-ir -vvv
[⠢] Compiling...
[⠘] Compiling 8 files with 0.8.17
[⠃] Solc 0.8.17 finished in 241.63s
Compiler run successful

Running 3 tests for test/foundry/FleekERC721/FleekERC721.deploy.t.sol:Test_FleekERC721_Deploy
[PASS] test_deployerShouldBeCollectionOwner() (gas: 13057)
[PASS] test_name() (gas: 9555)
[PASS] test_symbol() (gas: 10132)
Test result: ok. 3 passed; 0 failed; finished in 1.07ms

Running 4 tests for test/foundry/FleekERC721/FleekERC721.burn.t.sol:Test_FleekERC721_Burn
[PASS] testFail_burnAsController() (gas: 89768)
[PASS] testFail_burnAsNotOwner() (gas: 16173)
[PASS] testFail_burnInexistentToken() (gas: 11136)
[PASS] test_burn() (gas: 155720)
Test result: ok. 4 passed; 0 failed; finished in 1.14ms

Running 4 tests for test/foundry/FleekERC721/FleekERC721.tokenURI.t.sol:Test_FleekERC721_URI
[PASS] testFail_tokenURIForInexistentId() (gas: 8338)
[PASS] test_tokenURI() (gas: 2127002)
[PASS] test_tokenURIAfterUpdate() (gas: 2741131)
[PASS] test_tokenURIForDifferentAddresses() (gas: 5956173)
Test result: ok. 4 passed; 0 failed; finished in 20.86ms

Running 12 tests for test/foundry/FleekERC721/FleekERC721.ap.t.sol:Test_FleekERC721_AccessPoint
[PASS] test_addAccessPoint() (gas: 88723)
[PASS] test_appAccessPoints() (gas: 253848)
[PASS] test_cannotAddAccessPointToUnexistentToken() (gas: 11938)
[PASS] test_cannotDecreaseAccessPointScoreToMinusOne() (gas: 124159)
[PASS] test_cannotRemoveUnexistentAccessPoint() (gas: 11931)
[PASS] test_cannotTwiceRemoveAccessPoint() (gas: 78374)
[PASS] test_decreaseAccessPointScore() (gas: 185379)
[PASS] test_getAccessPointJSON() (gas: 118474)
[PASS] test_increateAccessPointScore() (gas: 205035)
[PASS] test_isAccessPointNameVerified() (gas: 104921)
[PASS] test_removeAccessPoint() (gas: 78484)
[PASS] test_setAccessPointVerifiesWithCorrectRole() (gas: 232631)
Test result: ok. 12 passed; 0 failed; finished in 33.61ms

Running 4 tests for test/foundry/FleekERC721/FleekERC721.mint.t.sol:Test_FleekERC721_Mint
[PASS] test_balanceOfDeployerAfterAndBeforeMinting() (gas: 903324)
[PASS] test_mint() (gas: 901744)
[PASS] test_mintTwoTokensForTwoAddresses() (gas: 2099458)
[PASS] test_mintingMintedToken() (gas: 1745849)
Test result: ok. 4 passed; 0 failed; finished in 33.88ms

Running 14 tests for test/foundry/FleekERC721/FleekERC721.access.t.sol:Test_FleekERC721_AccessControl
[PASS] test_grantAndRevokeCollectionRole() (gas: 93626)
[PASS] test_grantAndRevokeTokenRole() (gas: 106454)
[PASS] test_mint() (gas: 947977)
[PASS] test_setTokenBuild() (gas: 185805)
[PASS] test_setTokenColor() (gas: 68499)
[PASS] test_setTokenDescription() (gas: 75587)
[PASS] test_setTokenENS() (gas: 76319)
[PASS] test_setTokenExternalURL() (gas: 78532)
[PASS] test_setTokenLogo() (gas: 142805)
[PASS] test_setTokenLogoAndColor() (gas: 150785)
[PASS] test_setTokenName() (gas: 77794)
[PASS] test_setUp() (gas: 107582)
[PASS] test_testBurn() (gas: 189731)
[PASS] test_tokenURI() (gas: 9902529)
Test result: ok. 14 passed; 0 failed; finished in 33.77ms
✨  Done in 242.37s.
```

### CODE REVIEW

File: test/foundry/FleekERC721/FleekERC721.ap.t.sol [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        assertEq(current, expectedJSON);
+    }
+
+    function setUp() public {
+        baseSetUp();
+        tokenId = mintDefault(deployer);
+    }
+
+    function test_addAccessPoint() public {
+        CuT.addAccessPoint(tokenId, "accesspoint.com");
```
Comment (jsonsivar) [inline test/foundry/FleekERC721/FleekERC721.ap.t.sol:33]: Should we expect by calling a read function on the token Id?

**Review by jsonsivar — APPROVED** [review]

File: package.json [hunk 2]
```diff
@@ -5,7 +5,7 @@
   "private": "false",
   "scripts": {
     "test": "hardhat test && forge test --via-ir",
-    "test:foundry": "forge test --via-ir",
+    "test:foundry": "forge test --via-ir -vvv",
```
[the remark this answers is not in the document]
Comment (zoruka) [reply package.json:8]: Yeah I think it makes sense in that way because `test:foundry` mainly will be run to debug, since using just `test` implies in not showing the full verbose, which would be run to run both foundry and hardhat tests.

Also a good note about that is that `-vvv` will show verbose for failing tests and `-vvvv` will show full verbose for all tests.

File: test/foundry/FleekERC721/TestBase.sol [hunk 3]
```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.17;
+
+import "forge-std/Test.sol";
+import {FleekERC721} from "contracts/FleekERC721.sol";
+import {TestConstants} from "./Constants.sol";
+
+abstract contract Test_FleekERC721_Assertions is Test {
+    function expectRevertWithTokenRole() public {
```
[the remark this answers is not in the document]
Comment (zoruka) [reply test/foundry/FleekERC721/TestBase.sol:10]: I think it may be used by the others modules also.

File: test/foundry/FleekERC721/FleekERC721.access.t.sol [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+        assertFalse(CuT.hasTokenRole(tokenId, FleekAccessControl.Roles.Controller, anyAddress));
+    }
+
+    function test_grantAndRevokeCollectionRole() public {
+        address randomAddress = address(99);
+
+        // CollectionOwner
+        vm.startPrank(collectionOwner);
+        CuT.grantCollectionRole(FleekAccessControl.Roles.Controller, randomAddress);
+        CuT.revokeCollectionRole(FleekAccessControl.Roles.Controller, randomAddress);
```
[the remark this answers is not in the document]
Comment (zoruka) [reply test/foundry/FleekERC721/FleekERC721.access.t.sol:64]: Do you mean an assertion to check if `randomAddress` got the role?

File: test/foundry/FleekERC721/FleekERC721.access.t.sol [hunk 5]
```diff
@@ -0,0 +1,440 @@
+// SPDX-License-Identifier: MIT
```
[the remark this answers is not in the document]
Comment (zoruka) [reply test/foundry/FleekERC721/FleekERC721.access.t.sol:1]: I did follow this pattern from https://book.getfoundry.sh/tutorials/best-practices#general-test-guidance, but I think is up to us if we want to have it or not.

File: test/foundry/FleekERC721/FleekERC721.ap.t.sol (hunk 1 above)
Comment (zoruka) [reply test/foundry/FleekERC721/FleekERC721.ap.t.sol:33]: Do you mean by calling `tokenURI(tokenId)`? By now it would return a Base64, what I think is hard to parse in Solidity. What would help here is adding the function that returns the Solidity object, as it is in #93. Maybe we can add this assertion after implementing this function.

File: test/foundry/FleekERC721/FleekERC721.access.t.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (zoruka) [reply test/foundry/FleekERC721/FleekERC721.access.t.sol:65]: Added.

File: test/foundry/FleekERC721/FleekERC721.access.t.sol [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+        expectRevertWithTokenRole();
+        CuT.grantTokenRole(tokenId, FleekAccessControl.Roles.Controller, randomAddress);
+        expectRevertWithTokenRole();
+        CuT.revokeTokenRole(tokenId, FleekAccessControl.Roles.Controller, randomAddress);
+        vm.stopPrank();
+
+        // TokenOwner
+        vm.startPrank(tokenOwner);
+        CuT.grantTokenRole(tokenId, FleekAccessControl.Roles.Controller, randomAddress);
+        CuT.revokeTokenRole(tokenId, FleekAccessControl.Roles.Controller, randomAddress);
```
[the remark this answers is not in the document]
Comment (zoruka) [reply test/foundry/FleekERC721/FleekERC721.access.t.sol:125]: Added.

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (EmperorOrokuSaki) [reply package.json:8]: I agree with going verbose on failing tests.

File: test/foundry/FleekERC721/FleekERC721.access.t.sol [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+        vm.startPrank(anyAddress);
+        expectRevertWithCollectionRole();
+        mintDefault(randomAddress);
+        vm.stopPrank();
+    }
+
+    function test_tokenURI() public {
+        // ColletionOwner
+        vm.prank(collectionOwner);
+        CuT.tokenURI(tokenId);
```
Comment (EmperorOrokuSaki) [inline test/foundry/FleekERC721/FleekERC721.access.t.sol:182]: I don't think it's necessary to check all of these addresses if  `anyAddress` works.

File: test/foundry/FleekERC721/FleekERC721.ap.t.sol [hunk 8]
```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: MIT
```
Comment (EmperorOrokuSaki) [inline test/foundry/FleekERC721/FleekERC721.ap.t.sol:1]: can we rename this file to `FleekERC721.accessPoint.t.sol`? It's more descriptive.

Comment (EmperorOrokuSaki) [reply test/foundry/FleekERC721/FleekERC721.ap.t.sol:1]: Same for `FleekERC721.access.t.sol`.
Maybe we can come up with a better name.

File: test/foundry/FleekERC721/FleekERC721.ap.t.sol (hunk 1 above)
Comment (EmperorOrokuSaki) [reply test/foundry/FleekERC721/FleekERC721.ap.t.sol:33]: For now, we can just check if the access point is actually added and if it exists. To check the details, we can expose the function that is suggested in Issue #93 later.

File: test/foundry/FleekERC721/FleekERC721.ap.t.sol [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+    function test_addAccessPoint() public {
+        CuT.addAccessPoint(tokenId, "accesspoint.com");
+    }
+
+    function test_getAccessPointJSON() public {
+        string memory accessPointName = "accesspoint.com";
+        CuT.addAccessPoint(tokenId, accessPointName);
+
+        assertAccessPointJSON(accessPointName, "0", "0", "false", "false", deployer);
```
Comment (EmperorOrokuSaki) [inline test/foundry/FleekERC721/FleekERC721.ap.t.sol:40]: This is what I meant by my comment earlier about checking the existence of the access point. Maybe `test_addAccessPoint` and `test_getAccessPointJSON` can be merged into one test case since they share a similar purpose and the latter complements the former.

File: test/foundry/FleekERC721/FleekERC721.ap.t.sol [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    function test_isAccessPointNameVerified() public {
+        string memory accessPointName = "accesspoint.com";
+        CuT.addAccessPoint(tokenId, accessPointName);
+        assertFalse(CuT.isAccessPointNameVerified(accessPointName));
+        CuT.setAccessPointNameVerify(accessPointName, true);
+    }
+
+    function test_increateAccessPointScore() public {
```
Comment (EmperorOrokuSaki) [inline test/foundry/FleekERC721/FleekERC721.ap.t.sol:73]:
```suggestion
    function test_increaseAccessPointScore() public {
```

File: test/foundry/FleekERC721/FleekERC721.ap.t.sol [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    function test_removeAccessPoint() public {
+        string memory accessPointName = "accesspoint.com";
+        CuT.addAccessPoint(tokenId, accessPointName);
+        CuT.removeAccessPoint(accessPointName);
+
+        expectRevertWithInvalidAP();
+        CuT.getAccessPointJSON(accessPointName);
+    }
+
+    function test_cannotRemoveUnexistentAccessPoint() public {
```
Comment (EmperorOrokuSaki) [inline test/foundry/FleekERC721/FleekERC721.ap.t.sol:52]:
```suggestion
    function test_cannotRemoveNonexistentAccessPoint() public {
```

File: test/foundry/FleekERC721/FleekERC721.ap.t.sol [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+
+    function test_appAccessPoints() public {
+        CuT.addAccessPoint(tokenId, "accesspoint1.com");
+        CuT.addAccessPoint(tokenId, "accesspoint2.com");
+        CuT.addAccessPoint(tokenId, "accesspoint3.com");
+
+        string[] memory accessPoints = CuT.appAccessPoints(tokenId);
+        assertEq(accessPoints[0], "accesspoint1.com");
+        assertEq(accessPoints[1], "accesspoint2.com");
+        assertEq(accessPoints[2], "accesspoint3.com");
```
Comment (EmperorOrokuSaki) [inline test/foundry/FleekERC721/FleekERC721.ap.t.sol:113]: I suggest we add an assertion here for the length of `accessPoints` to confirm it only contains three values.

File: test/foundry/FleekERC721/FleekERC721.ap.t.sol [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    function test_cannotAddAccessPointToUnexistentToken() public {
```
Comment (EmperorOrokuSaki) [inline test/foundry/FleekERC721/FleekERC721.ap.t.sol:116]:
```suggestion
    function test_cannotAddAccessPointToNonexistentToken() public {
```

**Review by EmperorOrokuSaki — CHANGES_REQUESTED** [review]
Great work! All foundry tests are much more readable now. I love this.

I have left a few suggestions and requested some changes on this pull request. Let me know if you agree or not with them.

By the way, is there a reason for test names like `test_cannot`? I feel like going camelCase with all test names is better than having an underline between the characters.

I suggest we follow Foundry's docs and rename those tests to `testCannot`.

And also, I recommend using `testCannot` for tests that are going to fail in the end. This way, we can be sure of the reason. (Also from Foundry's guides) 

Line 32 of the `FleekERC721.burn.t.sol` file can be an example of this.

File: test/foundry/FleekERC721/FleekERC721.ap.t.sol (hunk 9 above)
Comment (zoruka) [reply test/foundry/FleekERC721/FleekERC721.ap.t.sol:40]: I will be removing the `test_addAccessPoint`

**Review by jsonsivar — APPROVED** [review]
LGTM.  There are 2 comments still as unresolved but looks like the code change for that is in so @EmperorOrokuSaki kindly mark those as resolved and then we should be good to merge if you are also good with it.
